src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,241 @@
+		bool *columnNulls = slot->tts_isnull;
+		FmgrInfo *columnOutputFunctions = copyDest->columnOutputFunctions;
+		CopyCoercionData *columnCoercionPaths = copyDest->columnCoercionPaths;
+
+		AppendCopyRowData(columnValues, columnNulls, copyDest->tupleDescriptor,
+						  copyDest->copyOutState, columnOutputFunctions,
+						  columnCoercionPaths);
+	}
+}
+
+
+/*
+ * ShouldSendCopyNow returns true if the given buffer size exceeds the
+ * local copy buffer size threshold.
+ */
+static bool
+ShouldSendCopyNow(StringInfo buffer)
+{
+	return buffer->len > LOCAL_COPY_BUFFER_SIZE;
+}
+
+
+/*
+ * DoLocalCopy finds the shard table from the distributed relation id, and copies the given
+ * buffer into the shard.
+ */
+static void
+DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId, CopyStmt *copyStatement,
+			bool isEndOfCopy)
+{
+	localCopyBuffer = buffer;
+
+	Oid shardOid = GetShardLocalTableOid(relationId, shardId);
+	Relation shard = heap_open(shardOid, RowExclusiveLock);
+	Relation copiedShard = CreateCopiedShard(copyStatement->relation, shard);
+	ParseState *pState = make_parsestate(NULL);
+
+	/* p_rtable of pState is set so that we can check constraints. */
+	pState->p_rtable = CreateRangeTable(copiedShard, ACL_INSERT);
+
+	CopyState cstate = BeginCopyFrom(pState, copiedShard, NULL, false,
+									 ReadFromLocalBufferCallback,
+									 copyStatement->attlist, copyStatement->options);
+	CopyFrom(cstate);
+	EndCopyFrom(cstate);
+
+	heap_close(shard, NoLock);
+	free_parsestate(pState);
+	FreeStringInfo(buffer);
+	if (!isEndOfCopy)
+	{
+		buffer = makeStringInfo();
+	}
+}
+
+
+/*
+ * ShouldAddBinaryHeaders returns true if the given buffer
+ * is empty and the format is binary.
+ */
+static bool
+ShouldAddBinaryHeaders(StringInfo buffer, bool isBinary)
+{
+	if (!isBinary)
+	{
+		return false;
+	}
+	return buffer->len == 0;
+}
+
+
+/*
+ * CreateCopiedShard clones deep copies the necessary fields of the given
+ * relation.
+ */
+Relation
+CreateCopiedShard(RangeVar *distributedRel, Relation shard)
+{
+	TupleDesc tupleDescriptor = RelationGetDescr(shard);
+
+	Relation copiedDistributedRelation = (Relation) palloc(sizeof(RelationData));
+	Form_pg_class copiedDistributedRelationTuple = (Form_pg_class) palloc(
+		CLASS_TUPLE_SIZE);
+
+	*copiedDistributedRelation = *shard;
+	*copiedDistributedRelationTuple = *shard->rd_rel;
+
+	copiedDistributedRelation->rd_rel = copiedDistributedRelationTuple;
+	copiedDistributedRelation->rd_att = CreateTupleDescCopyConstr(tupleDescriptor);
+
+	Oid tableId = RangeVarGetRelid(distributedRel, NoLock, false);
+
+	/*
+	 * BeginCopyFrom opens all partitions of given partitioned table with relation_open
+	 * and it expects its caller to close those relations. We do not have direct access
+	 * to opened relations, thus we are changing relkind of partitioned tables so that
+	 * Postgres will treat those tables as regular relations and will not open its
+	 * partitions.
+	 */
+	if (PartitionedTable(tableId))
```

Comment:
I kind of copy pasted this, do we really need this?

Reply:
well it seems that we should not have this as we wont be copying into the right partition, removed it.